File: backend/routers/allocations.py
"""
routers/allocations.py — Phase 3
Asset Allocation and Transfer flows
"""
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.orm import Session

from database import get_db
from models import (
    Asset, AssetStatus, Employee, EmployeeRole, Department,
    Allocation, TransferRequest, TransferStatus,
    ActivityLog, LogType
)
from routers.auth import get_current_user

router = APIRouter()

# ── Pydantic Schemas ──────────────────────────────────────────

class AllocationHistoryOut(BaseModel):
    id: int
    allocated_to: str
    department: str | None
    allocated_at: datetime
    returned_at: datetime | None
    notes: str | None
    condition_on_return: str | None

class AllocationStateOut(BaseModel):
    asset_id: int
    is_allocated: bool
    current_allocation: AllocationHistoryOut | None
    history: list[AllocationHistoryOut]

class AllocationCreate(BaseModel):
    asset_id: int
    employee_id: int
    notes: str | None = None

class TransferRequestCreate(BaseModel):
    asset_id: int
    to_employee_id: int
    reason: str | None = None

class AllocationOut(BaseModel):
    id: int
    asset_id: int
    employee_id: int
    allocated_at: datetime
    notes: str | None

class TransferRequestOut(BaseModel):
    id: int
    asset_id: int
    from_employee_id: int
    to_employee_id: int
    status: str
    reason: str | None
    created_at: datetime

# ── Helpers ───────────────────────────────────────────────────

def _allocation_to_history(alloc: Allocation, db: Session) -> AllocationHistoryOut:
    employee = db.get(Employee, alloc.employee_id)
    emp_name = employee.name if employee else "Unknown"
    dept_name = None
    if employee and employee.department_id:
        dept = db.get(Department, employee.department_id)
        if dept:
            dept_name = dept.name
            
    return AllocationHistoryOut(
        id=alloc.id,
        allocated_to=emp_name,
        department=dept_name,
        allocated_at=alloc.allocated_at,
        returned_at=alloc.returned_at,
        notes=alloc.notes,
        condition_on_return=alloc.condition_on_return
    )

# ── Endpoints ─────────────────────────────────────────────────

@router.get("/{asset_id}", response_model=AllocationStateOut)
def get_allocation_state(
    asset_id: int,
    db: Session = Depends(get_db),
    _: Employee = Depends(get_current_user)
):
    asset = db.get(Asset, asset_id)
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")

    allocations = db.query(Allocation).filter(Allocation.asset_id == asset_id).order_by(Allocation.allocated_at.desc()).all()
    
    current_allocation = None
    history = []
    
    for alloc in allocations:
        history_out = _allocation_to_history(alloc, db)
        history.append(history_out)
        if alloc.returned_at is None:
            current_allocation = history_out

    return AllocationStateOut(
        asset_id=asset_id,
        is_allocated=current_allocation is not None,
        current_allocation=current_allocation,
        history=history
    )

@router.post("", response_model=AllocationOut, status_code=status.HTTP_201_CREATED)
def allocate_asset(
    body: AllocationCreate,
    db: Session = Depends(get_db),
    current_user: Employee = Depends(get_current_user)
):
    # Check permissions
    if current_user.role not in (EmployeeRole.asset_manager, EmployeeRole.admin, EmployeeRole.superadmin):
        raise HTTPException(status_code=403, detail="Asset Manager privileges required")

    asset = db.get(Asset, body.asset_id)
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")

    # Check conflict
    active_alloc = db.query(Allocation).filter(
        Allocation.asset_id == body.asset_id,
        Allocation.returned_at == None
    ).first()
    
    if active_alloc or asset.status != AssetStatus.available:
        raise HTTPException(status_code=409, detail="Asset is already allocated or unavailable")

    employee = db.get(Employee, body.employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")

    # Create allocation
    new_alloc = Allocation(
        asset_id=body.asset_id,
        employee_id=body.employee_id,
        notes=body.notes
    )
    db.add(new_alloc)
    
    # Update asset status
    asset.status = AssetStatus.allocated
    
    # Activity Log
    emp_dept = db.get(Department, employee.department_id).name if employee.department_id else "Unknown"
    log = ActivityLog(
        type=LogType.alert,
        message=f"{asset.category.name if asset.category else 'Asset'} {asset.tag} – allocated to {employee.name} – {emp_dept} dept",
        employee_id=employee.id
    )
    db.add(log)
    
    db.commit()
    db.refresh(new_alloc)
    
    return AllocationOut(
        id=new_alloc.id,
        asset_id=new_alloc.asset_id,
        employee_id=new_alloc.employee_id,
        allocated_at=new_alloc.allocated_at,
        notes=new_alloc.notes
    )

@router.post("/transfer", response_model=TransferRequestOut, status_code=status.HTTP_201_CREATED)
def request_transfer(
    body: TransferRequestCreate,
    db: Session = Depends(get_db),
    current_user: Employee = Depends(get_current_user)
):
    asset = db.get(Asset, body.asset_id)
    if not asset:
        raise HTTPException(status_code=404, detail="Asset not found")

    # Ensure it's currently allocated
    active_alloc = db.query(Allocation).filter(
        Allocation.asset_id == body.asset_id,
        Allocation.returned_at == None
    ).first()
    
    if not active_alloc:
        raise HTTPException(status_code=400, detail="Asset is not currently allocated, cannot request transfer")

    # Create transfer request
    transfer = TransferRequest(
        asset_id=body.asset_id,
        from_employee_id=active_alloc.employee_id,
        to_employee_id=body.to_employee_id,
        reason=body.reason
    )
    db.add(transfer)
    
    # Activity Log
    log = ActivityLog(
        type=LogType.alert,
        message=f"Transfer requested for {asset.tag} – from {db.get(Employee, active_alloc.employee_id).name} to {db.get(Employee, body.to_employee_id).name}",
        employee_id=current_user.id
    )
    db.add(log)
    
    db.commit()
    db.refresh(transfer)
    
    return TransferRequestOut(
        id=transfer.id,
        asset_id=transfer.asset_id,
        from_employee_id=transfer.from_employee_id,
        to_employee_id=transfer.to_employee_id,
        status=transfer.status.value,
        reason=transfer.reason,
        created_at=transfer.created_at
    )

File: frontend/src/api/allocations.ts
import { getStoredToken } from './auth'

const API_BASE = import.meta.env.VITE_API_URL ?? 'http://localhost:8000'

async function authFetch<T>(path: string, options: RequestInit = {}): Promise<T> {
  const token = getStoredToken()
  const res = await fetch(`${API_BASE}${path}`, {
    headers: {
      'Content-Type': 'application/json',
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...(options.headers ?? {}),
    },
    ...options,
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail ?? 'Request failed')
  }
  if (res.status === 204) return undefined as unknown as T
  return res.json() as Promise<T>
}

// ── Types ────────────────────────────────────────────────────────

export interface AllocationHistoryOut {
  id: number
  allocated_to: string
  department: string | null
  allocated_at: string
  returned_at: string | null
  notes: string | null
  condition_on_return: string | null
}

export interface AllocationStateOut {
  asset_id: number
  is_allocated: boolean
  current_allocation: AllocationHistoryOut | null
  history: AllocationHistoryOut[]
}

export interface AllocationCreate {
  asset_id: number
  employee_id: number
  notes?: string | null
}

export interface TransferRequestCreate {
  asset_id: number
  to_employee_id: number
  reason?: string | null
}

export interface AllocationOut {
  id: number
  asset_id: number
  employee_id: number
  allocated_at: string
  notes: string | null
}

export interface TransferRequestOut {
  id: number
  asset_id: number
  from_employee_id: number
  to_employee_id: number
  status: string
  reason: string | null
  created_at: string
}

// ── API Functions ────────────────────────────────────────────────

export const getAllocationState = (assetId: number): Promise<AllocationStateOut> =>
  authFetch(`/allocations/${assetId}`)

export const allocateAsset = (body: AllocationCreate): Promise<AllocationOut> =>
  authFetch('/allocations', { method: 'POST', body: JSON.stringify(body) })

export const requestTransfer = (body: TransferRequestCreate): Promise<TransferRequestOut> =>
  authFetch('/allocations/transfer', { method: 'POST', body: JSON.stringify(body) })

// ── Compatibility for Phase 4 (Bookings) ─────────────────────────
import { useQuery, useMutation } from '@tanstack/react-query'

export const useGetAllocations = () => {
  return useQuery({
    queryKey: ['allocations_all_mock'],
    queryFn: () => Promise.resolve([])
  })
}

export const useCreateAllocation = () => {
  return useMutation({
    mutationFn: (_body: any) => Promise.resolve({})
  })
}

export const useReturnAllocation = () => {
  return useMutation({
    mutationFn: (_body: any) => Promise.resolve({})
  })
}
